onnxruntime/core/session/inference_session.h

```diff
@@ -167,7 +167,7 @@ class InferenceSession {
     * Calling this API is optional in which case onnxruntime will use its internal CPU execution provider.
     * @return OK if success.
     */
-  common::Status RegisterExecutionProvider(std::unique_ptr<IExecutionProvider> p_exec_provider) ORT_MUST_USE_RESULT;
```

Comment:
(Apologies for commenting on a draft PR, but wanted to make sure I didn't forget this...)

Would it make sense to keep the `std::unique_ptr<IExecutionProvider>` method for backwards compatibility, and "upgrade" it to a `shared_ptr` internally?

Reply:
unfortunately this is not doable... 
when we get the CUDA EP from the cache, it's already held in a shared_ptr, so the interface also need to be updated to accept a shared_ptr

Reply:
I was thinking of something different... for current C++ users whose code currently passes a unique_ptr, I think you could retain the old method as well, and its implementation could simply "forward" the argument to the new shared_ptr overload.

[EDIT: After writing this, I realize that it's not necessary because there's already [an implict shared_ptr constructor](https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr) that takes a `std::unique_ptr<T>&&`. Never mind!]